### PR TITLE
fix: `shots=0` measurements with trainable params

### LINKED ISSUE #129: SV1 calculates only the gradient if parameters are explicitly marked `requires_grad=True`
**Describe the bug**
If parameters passed into a qnode are PennyLane tensors explicitly marked with `requires_grad=True`, BraketAwsQubitDevice with SV1 will only calculate measurements through the AdjointGradient result type, which in turn only supports expectation values. As another consequence, attempting to run circuits with other measurements (like var), with more than one observable, or no observables at all (like the Probability measurement) will fail altogether.

**To reproduce**
```python
In [1]: import pennylane as qml
   ...: from pennylane import numpy as qnp

In [2]: device_arn = "arn:aws:braket:::device/quantum-simulator/amazon/sv1"
   ...: dev = qml.device("braket.aws.qubit", device_arn=device_arn, wires=2, shots=0)

In [3]: weights = qnp.array([0.1, 0.2, 0.3], requires_grad=True)  # trainable

In [4]: weights2 = [0.1, 0.2, 0.3]  # not trainable

In [5]: @qml.qnode(dev)
   ...: def circuit1(a, weights):
   ...:     qml.RZ(a, wires=0)
   ...:     qml.RY(a, wires=1)
   ...:     qml.RX(weights[0], wires=0)
   ...:     qml.RY(weights[1], wires=1)
   ...:     qml.CNOT(wires=[0, 1])
   ...:     qml.RZ(3.14, wires=1)
   ...:     qml.RX(weights[0], wires=1)
   ...:     qml.RX(a, wires=1)
   ...:     return qml.expval(qml.PauliX(wires=0) @ qml.PauliY(wires=1))
   ...: 

In [6]: circuit1(3, weights)  # trainable
   ...: print(dev.circuit)
T  : |  0  |   1   |2|   3    |   4   |  5  |    Result Types    |
                                                                  
q0 : -Rz(3)-Rx(p_2)-C------------------------AdjointGradient(X@Y)-
                    |                        |                    
q1 : -Ry(3)-Ry(p_3)-X-Rz(3.14)-Rx(p_5)-Rx(3)-AdjointGradient(X@Y)-

T  : |  0  |   1   |2|   3    |   4   |  5  |    Result Types    |

Unassigned parameters: [p_2, p_3, p_5].

In [7]: circuit1(3, weights2)  # not trainable
   ...: print(dev.circuit)
T  : |  0  |   1    |2|   3    |   4    |  5  |  Result Types  |
                                                                
q0 : -Rz(3)-Rx(0.10)-C-------------------------Expectation(X@Y)-
                     |                         |                
q1 : -Ry(3)-Ry(0.20)-X-Rz(3.14)-Rx(0.10)-Rx(3)-Expectation(X@Y)-

T  : |  0  |   1    |2|   3    |   4    |  5  |  Result Types  |
```
**Expected behavior**
The device should calculate any requested measurement rather than just expectation.

**Screenshots or logs**
If applicable, add screenshots or logs to help explain your problem.

**System information**
A description of your system. Please provide:
- **Amazon Braket Python PennyLane Plugin version**: v1.10.0 (and up)
- **Amazon Braket Python SDK version**:
- **Amazon Braket Python Schemas version**:
- **Amazon Braket Python Default Simulator version**:
- **Python version**:

**Workaround**
Please use versions of the plugin <1.10.0 if you want to use tensors with `requires_grad=True`.

### PR DESCRIPTION
**Pull request by speller26**
Fixes #129 by establishing a code path for running circuits with trainable parameters and either variance measurements or multiple observables when shots=0.

*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/unit_tests/test_braket_device.py [hunk 1]
```diff
@@ -1512,15 +1588,15 @@ def _bad_aws_device(properties_mock, session_mock, wires, **kwargs):
     )
 
 
-def get_test_result_object(result_types=[], source="qubit[2] q; cnot q[0], q[1]; measure q;"):
```
[the remark this answers is not in the document]
Comment (speller26) [reply test/unit_tests/test_braket_device.py:1515]: I just really don't like shadowing scope, since `result_types` is already imported

File: src/braket/pennylane_plugin/braket_device.py [hunk 2]
```diff
@@ -149,7 +151,7 @@ def task(self) -> QuantumTask:
         """QuantumTask: The task corresponding to the last run circuit."""
         return self._task
 
-    def _pl_to_braket_circuit(self, circuit: CircuitGraph, compute_gradient=False, **run_kwargs):
+    def _pl_to_braket_circuit(self, circuit: QuantumTape, compute_gradient=False, **run_kwargs):
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:154]: CircuitGraph is a misnomer here; the type that's actually being passed by PL core is a QuantumTape.

File: src/braket/pennylane_plugin/braket_device.py [hunk 3]
```diff
@@ -544,17 +546,26 @@ def execute_and_gradients(self, circuits, **kwargs):
...
         res = []
         jacs = []
         for circuit in circuits:
+            observables: List[Observable] = circuit.observables
             if not circuit.trainable_params:
                 new_res = self.execute(circuit, compute_gradient=False)
                 # don't bother computing a gradient when there aren't any trainable parameters.
                 new_jac = np.tensor([])
+            elif len(observables) != 1 or observables[0].return_type != Expectation:
+                gradient_circuits, post_processing_fn = param_shift(circuit)
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:563]: Yep, just as it did before AG was introduced

File: test/unit_tests/test_braket_device.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacofeld) [reply test/unit_tests/test_braket_device.py:1515]: Ahhh I see, makes sense. I must not have noticed at the time I wrote this either haha

**Review by jacofeld — APPROVED** [review]
